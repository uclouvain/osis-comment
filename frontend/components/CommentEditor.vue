<!--
  -
  - OSIS stands for Open Student Information System. It's an application
  - designed to manage the core business of higher education institutions,
  - such as universities, faculties, institutes and professional schools.
  - The core business involves the administration of students, teachers,
  - courses, programs and so on.
  -
  - Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU General Public License as published by
  - the Free Software Foundation, either version 3 of the License, or
  - (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU General Public License for more details.
  -
  - A copy of this license - GNU General Public License - is available
  - at the root of the source code of this program.  If not,
  - see http://www.gnu.org/licenses/.
  -
  -->
<template>
  <div>
    <textarea
        v-if="!richTextConfig"
        v-model="value"
        class="form-control"
        :placeholder="$t('entry.placeholder')"
    />
    <ckeditor
        v-else
        v-model="value"
        :config="richTextConfig"
    />
    <div class="text-right">
      <button
          class="btn btn-sm btn-success"
          @click="$emit('submit', value)"
      >
        {{ $t('entry.submit_comment') }}
      </button>
      <button
          class="btn btn-sm btn-default"
          @click="$emit('cancel')"
      >
        {{ $t('entry.cancel_edit') }}
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import component from './ckeditor.js';
import {defineComponent} from 'vue';
import type {Component} from 'vue';

export default defineComponent({
  name: 'CommentEditor',
  components: {
    ckeditor: (component as Component),
  },
  props: {
    richTextConfig: {
      type: Object,
      default: () => undefined,
    },
    defaultValue: {
      type: String,
      default: '',
    },
  },
  emits: ['cancel', 'submit'],
  data: function () {
    return {
      value: this.defaultValue,
    };
  },
});
</script>
