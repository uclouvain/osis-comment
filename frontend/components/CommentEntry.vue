<!--
  -
  - OSIS stands for Open Student Information System. It's an application
  - designed to manage the core business of higher education institutions,
  - such as universities, faculties, institutes and professional schools.
  - The core business involves the administration of students, teachers,
  - courses, programs and so on.
  -
  - Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU General Public License as published by
  - the Free Software Foundation, either version 3 of the License, or
  - (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU General Public License for more details.
  -
  - A copy of this license - GNU General Public License - is available
  - at the root of the source code of this program.  If not,
  - see http://www.gnu.org/licenses/.
  -
  -->

<template>
  <div class="panel-body clearfix">
    <div class="comment-authoring">
      <strong>{{ entry.author || $t('entry.anonymous') }}</strong>
      {{
        $t('entry.authored_date', {
          date: entry.created_at.toLocaleDateString(),
          time: entry.created_at.toLocaleTimeString(),
        })
      }}
    </div>
    <template v-if="!isEditing">
      <div class="pull-right comment-actions">
        <button
            v-if="typeof entry.links.edit === 'string'"
            class="btn btn-sm btn-default"
            :title="$t('entry.edit_comment')"
            @click="isEditing = true"
        >
          <i class="fas fa-pencil" />
        </button>
        <button
            v-if="typeof entry.links.delete === 'string'"
            class="btn btn-sm btn-danger"
            :title="$t('entry.delete_comment')"
            @click="$emit('delete', entry.links.delete)"
        >
          <i class="fas fa-trash-alt" />
        </button>
      </div>
      <div
          v-if="!richTextConfig"
          class="comment-content"
      >
        <!-- eslint-disable vue/no-v-html -->
        <p v-html="entry.comment.replace(/\n/g, '<br/>')" />
      </div>
      <!-- eslint-disable vue/no-v-html -->
      <div
          v-else
          class="comment-content"
          v-html="entry.comment"
      />
    </template>
    <CommentEditor
        v-else
        :default-value="entry.comment"
        :rich-text-config="richTextConfig"
        @submit="(value) => $emit('edit', entry.links.edit, value)"
        @cancel="isEditing = false"
    />
  </div>
</template>

<script lang="ts">
import {Entry} from '../types';
import CommentEditor from './CommentEditor.vue';
import {defineComponent} from "vue";

export default defineComponent({
  name: 'CommentEntry',
  components: {CommentEditor},
  props: {
    richTextConfig: {
      type: Object,
      default: () => undefined,
    },
    entry: {
      type: Entry,
      required: true,
    },
  },
  emits: ['edit', 'delete'],
  data: function () {
    return {
      isEditing: false,
    };
  },
});
</script>

<style>
.comment-authoring {
  background: #dddddd;
  padding: .5em 1em;
}

.comment-content {
  padding: 1em 0;
}

.comment-actions {
  padding: 4px 0;
}
</style>
