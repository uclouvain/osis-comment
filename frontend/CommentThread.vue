<!--
  -
  - OSIS stands for Open Student Information System. It's an application
  - designed to manage the core business of higher education institutions,
  - such as universities, faculties, institutes and professional schools.
  - The core business involves the administration of students, teachers,
  - courses, programs and so on.
  -
  - Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU General Public License as published by
  - the Free Software Foundation, either version 3 of the License, or
  - (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU General Public License for more details.
  -
  - A copy of this license - GNU General Public License - is available
  - at the root of the source code of this program.  If not,
  - see http://www.gnu.org/licenses/.
  -
  -->

<template>
  <div
      class="comment-thread panel"
      :class="`panel-${panelClass}`"
  >
    <div class="panel-heading clearfix">
      {{ headerTitle || $t('thread.title') }}
      <template v-if="total != null">
        ({{ total }})
      </template>
      <button
          class="btn btn-default btn-sm pull-right"
          @click="changeSort()"
      >
        {{ $t('thread.sort') }}
        <i
            class="fas"
            :class="isAscendedSort ? 'fa-sort-up' : 'fa-sort-down'"
        />
      </button>
    </div>
    <span
        v-if="loading"
        class="spinner"
    />
    <div
        v-else-if="error"
        class="panel-body text-danger"
    >
      {{ error }}
    </div>
    <CommentEntry
        v-for="entry in entries"
        v-else
        :key="entry.uuid"
        :entry="entry"
        :rich-text-config="richTextConfig"
        @edit="editEntry"
        @delete="deleteEntry"
    />
    <div
        v-if="nextPage || previousPage || createUrl"
        class="panel-body"
    >
      <button
          v-if="previousPage"
          class="btn btn-sm btn-default"
          @click="previousPage && (currentUrl = previousPage); loadEntries()"
      >
        <i class="fas fa-chevron-left" />
        {{ $t('thread.previous') }}
      </button>
      <button
          v-if="nextPage"
          class="btn btn-sm btn-default"
          @click="nextPage && (currentUrl = nextPage); loadEntries()"
      >
        {{ $t('thread.next') }}
        <i class="fas fa-chevron-right" />
      </button>
      <template v-if="createUrl">
        <button
            v-if="!isAdding"
            class="btn btn-sm btn-primary pull-right"
            @click="isAdding = !isAdding"
        >
          <i class="fas fa-plus" />
          {{ $t('thread.add_comment') }}
        </button>
        <CommentEditor
            v-else
            :rich-text-config="richTextConfig"
            @submit="createEntry"
            @cancel="isAdding = !isAdding"
        />
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import CommentEntry from './components/CommentEntry.vue';
import {Entry} from './types';
import type {EntriesResponse} from './interfaces';
import CommentEditor from './components/CommentEditor.vue';
import type {PropType} from "vue";
import EventBus from './event-bus';
import {defineComponent} from "vue";

export default defineComponent({
  name: 'CommentThread',
  components: {CommentEditor, CommentEntry},
  props: {
    url: {
      type: String,
      required: true,
    },
    tags: {
      type: Array as PropType<string[]>,
      default: () => [],
    },
    headerTitle: {
      type: String,
      default: '',
    },
    pageSize: {
      type: Number,
      default: 10,
    },
    defaultSort: {
      type: String,
      default: '-modified_at',
    },
    panelClass: {
      type: String,
      default: 'default',
    },
    richTextConfig: {
      type: Object,
      default: () => undefined,
    },
  },
  data() {
    const params = new URLSearchParams({
      limit: String(this.pageSize),
      sort: '-modified_at',
    });
    if (this.tags.length) {
      params.append('tags', this.tags.join(','));
    }
    const element = document.querySelector('[name=csrfmiddlewaretoken]');
    if (!(element instanceof HTMLInputElement)) {
      throw new Error('Please include {% csrf_token %} in your page.');
    }
    return {
      entries: [] as Entry[],
      error: '',
      total: null as number | null,
      createUrl: null as string | null,
      currentUrl: `${this.url}?${params.toString()}`,
      currentSort: '-modified_at',
      previousPage: null as string | null,
      nextPage: null as string | null,
      loading: true,
      isAdding: false,
      csrfToken: element.value,
    };
  },
  computed: {
    isAscendedSort: function (): boolean {
      return this.currentSort === 'modified_at';
    },
  },
  mounted() {
    void this.loadEntries();
  },
  methods: {
    async changeSort() {
      const url = new URL(window.location.origin + this.currentUrl);
      const params = new URLSearchParams(url.search);
      this.currentSort = this.isAscendedSort ? '-modified_at' : 'modified_at';
      params.set('sort', this.currentSort);
      this.currentUrl = `${url.pathname}?${params.toString()}`;
      await this.loadEntries();
    },
    async loadEntries() {
      const data = (await this.doRequest(this.currentUrl, {}, false)) as EntriesResponse;
      if (data) {
        this.createUrl = typeof data.create === 'string' ? data.create : null;
        this.previousPage = data.previous;
        this.nextPage = data.next;
        this.total = data.count;
        this.entries = data.results.map(r => new Entry(r));
      }
    },
    async createEntry(value: string) {
      await this.doRequest(this.createUrl as string, {
        method: 'POST',
        body: JSON.stringify({comment: value, tags: this.tags}),
      });
      this.isAdding = false;
    },
    async editEntry(url: string, value: string) {
      await this.doRequest(url, {
        method: 'PUT',
        body: JSON.stringify({comment: value}),
      });
    },
    async deleteEntry(url: string) {
      await this.doRequest(url, {method: 'DELETE'});
    },
    async doRequest(url: string, params: object, refresh = true) {
      this.loading = true;
      this.error = '';
      try {
        const response = await fetch(url, {
          mode: 'same-origin',
          headers: {
            'Content-Type': 'application/json;charset=utf-8',
            'X-CSRFToken': this.csrfToken,
          },
          ...params,
        });
        if (response.status >= 200 && response.status < 300) {
          if (refresh) {
            await this.loadEntries();
            EventBus.emit("REFRESH_COMMENTS_EVENT");
          }
          this.loading = false;
          return response.json();
        } else {
          this.error = response.statusText;
        }
      } catch (e) {
        this.error = (e as Error).message;
      }
      this.loading = false;
    },
  },
});
</script>

<style>
.comment-thread .panel-heading {
  line-height: 30px;
}
</style>
