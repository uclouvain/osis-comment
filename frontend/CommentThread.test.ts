/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {beforeEach, expect, test} from 'vitest';
import {flushPromises, mount} from '@vue/test-utils';
import CommentThread from "./CommentThread.vue";
import type {EntriesResponse} from "./interfaces";
import fetchMock from "fetch-mock";
import CommentEditor from "./components/CommentEditor.vue";
import CommentEntry from "./components/CommentEntry.vue";

const apiUrl = "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/";
const mockEntries: EntriesResponse = {
  results: [
    {
      uuid: "8784bf99-d4f6-4971-99cf-5227afbeca0f",
      comment: "Lorem ipsum dolor amet",
      created_at: "2023-03-03T11:37:05.493008",
      author: "John Doe",
      tags: [],
      extra_data: {},
      links: {
        "edit": "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f",
        "delete": "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f",
      },
    },
    {
      uuid: "8784bf99-d4f6-4971-99cf-5227afbeca0f",
      comment: "Consetetur sadipscing elitr, sed diam nonumy eirmod",
      created_at: "2023-03-02T11:37:05.493008",
      author: "John Doe",
      tags: [],
      extra_data: {},
      links: {
        "edit": {error: "Non"},
        "delete": {error: "Non"},
      },
    },
    {
      uuid: "8784bf99-d4f6-4971-99cf-5227afbeca0f",
      comment: "Et dolore magna aliquyam erat, sed diam voluptua.",
      created_at: "2023-03-01T11:37:05.493008",
      author: "Bob Smith",
      tags: [],
      extra_data: {},
      links: {
        "edit": "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f",
        "delete": "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f",
      },
    },
  ],
  count: 3,
  next: null,
  previous: null,
  create: apiUrl,
};

beforeEach(() => {
  const csrfToken = document.createElement('input');
  csrfToken.id = 'csrf-token';
  csrfToken.name = 'csrfmiddlewaretoken';
  document.body.appendChild(csrfToken);

  fetchMock.reset().mock('path:' + apiUrl, mockEntries);
});

test('thread wihout token', () => {
  (document.getElementById('csrf-token') as Element).remove();

  expect(() => mount(CommentThread, {
    props: {url: apiUrl},
  })).toThrowError('Please include {% csrf_token %} in your page.');
});

test('thread', async () => {
  expect(CommentThread).toBeTruthy();

  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });
  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
});


test('handle 404', async () => {
  fetchMock.reset().mock('path:' + apiUrl, 404);

  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();

  expect(wrapper.text()).toContain('Not Found');
});

test('handle network error', async () => {
  fetchMock.reset().mock('path:' + apiUrl, {throws: new Error('Some network error')});

  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();

  expect(wrapper.text()).toContain('Some network error');
});

test('change sort', async () => {
  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();
  expect(fetchMock.lastUrl()).toBe('/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/?limit=10&sort=-created');

  const sortButton = wrapper.get('.panel-heading .btn');
  expect(sortButton.get('i').classes()).toContain('fa-sort-down');

  await sortButton.trigger('click');
  expect(fetchMock.lastUrl()).toBe('/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/?limit=10&sort=created');
  expect(sortButton.get('i').classes()).toContain('fa-sort-up');

  await sortButton.trigger('click');
  expect(fetchMock.lastUrl()).toBe('/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/?limit=10&sort=-created');
  expect(sortButton.get('i').classes()).toContain('fa-sort-down');
});

test('handle pagination', async () => {
  const previousUrl = '/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/?limit=10&sort=-created&page=1';
  const nextUrl = '/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/?limit=10&sort=-created&page=3';
  fetchMock.reset().mock('path:' + apiUrl, {
    ...mockEntries,
    previous: previousUrl,
    next: nextUrl,
  });

  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();
  const [previousButton, nextButton] = wrapper.findAll('.panel-body:last-child .btn-default');
  await previousButton.trigger('click');
  expect(fetchMock.lastUrl()).toBe(previousUrl);

  await nextButton.trigger('click');
  expect(fetchMock.lastUrl()).toBe(nextUrl);
});

test('add comment', async () => {
  const wrapper = mount(CommentThread, {
    props: {url: apiUrl, tags: ['foo']},
  });

  await flushPromises();
  const addButton = wrapper.get('.btn-primary');
  expect(addButton.text()).toContain('thread.add_comment');
  expect(addButton.get('i').classes()).toContain('fa-plus');
  await addButton.trigger('click');

  const commentEditor = wrapper.findComponent(CommentEditor);
  expect(commentEditor.exists()).toBe(true);
  await commentEditor.get('textarea').setValue('Foobar');

  await commentEditor.get('button.btn-success').trigger('click');
  expect(commentEditor.emitted('submit')).toStrictEqual([['Foobar']]);
  await flushPromises();
  const postCall = (fetchMock.calls(true, 'POST')[0][1] as RequestInit).body;
  expect(postCall).toEqual(JSON.stringify({comment: "Foobar", tags: ['foo']}));
});

test('cancel adding comment', async () => {
  const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();
  await wrapper.get('.btn-primary').trigger('click');

  const commentEditor = wrapper.findComponent(CommentEditor);
  expect(commentEditor.exists()).toBe(true);
  await commentEditor.get('button.btn-default').trigger('click');
  expect(wrapper.findComponent(CommentEditor).exists()).toBe(false);
});

test('edit comment', async () => {
   const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();
  expect(wrapper.findAll('.comment-actions')).toHaveLength(3);

  const firstComment = wrapper.findAllComponents(CommentEntry)[0];

  const editButton = firstComment.get('.btn-default');
  expect(editButton.html()).toContain('entry.edit_comment');
  expect(editButton.get('i').classes()).toContain('fa-pencil');
  await editButton.trigger('click');

  const commentEditor = firstComment.findAllComponents(CommentEditor)[0];
  const textarea = commentEditor.find('textarea.form-control').element as HTMLInputElement;
  expect(textarea.value).toBe('Lorem ipsum dolor amet');
  expect(commentEditor.exists()).toBe(true);
  await commentEditor.get('textarea.form-control').setValue('Foobar');

  fetchMock.reset()
      .mock('path:/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f', 200)
      .mock('path:/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/', 200);
  await commentEditor.get('button.btn-success').trigger('click');
  await flushPromises();
  const postCall = (fetchMock.calls(true, 'PUT')[0][1] as RequestInit).body;
  expect(postCall).toEqual(JSON.stringify({comment: "Foobar"}));
});

test('delete comment', async () => {

   const wrapper = mount(CommentThread, {
    props: {url: apiUrl},
  });

  await flushPromises();
  expect(wrapper.findAll('.comment-actions')).toHaveLength(3);

  const firstComment = wrapper.findAllComponents(CommentEntry)[0];

  const editButton = firstComment.get('.btn-danger');
  expect(editButton.html()).toContain('entry.delete_comment');
  expect(editButton.get('i').classes()).toContain('fa-trash-alt');

  fetchMock.reset()
      .mock('path:/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f', 200)
      .mock('path:/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/', 200);
  await editButton.trigger('click');
  await flushPromises();
  expect(fetchMock.calls(true, 'DELETE')).toBeTruthy();
});
