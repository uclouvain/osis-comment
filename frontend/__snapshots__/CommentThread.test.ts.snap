// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`thread 1`] = `
"<div class=\\"comment-thread panel panel-default\\">
  <div class=\\"panel-heading clearfix\\">thread.title <button class=\\"btn btn-default btn-sm pull-right\\">thread.sort <i class=\\"fas fa-sort-down\\"></i></button></div>
  <div class=\\"panel-body clearfix\\">
    <div class=\\"comment-authoring\\"><strong>John Doe</strong> entry.authored_date</div>
    <div class=\\"pull-right comment-actions\\"><button class=\\"btn btn-sm btn-default\\" title=\\"entry.edit_comment\\"><i class=\\"fas fa-pencil\\"></i></button><button class=\\"btn btn-sm btn-danger\\" title=\\"entry.delete_comment\\"><i class=\\"fas fa-trash-alt\\"></i></button></div>
    <div class=\\"comment-content\\">
      <!-- eslint-disable vue/no-v-html -->
      <p>Lorem ipsum dolor amet</p>
    </div>
  </div>
  <div class=\\"panel-body clearfix\\">
    <div class=\\"comment-authoring\\"><strong>John Doe</strong> entry.authored_date</div>
    <div class=\\"pull-right comment-actions\\">
      <!--v-if-->
      <!--v-if-->
    </div>
    <div class=\\"comment-content\\">
      <!-- eslint-disable vue/no-v-html -->
      <p>Consetetur sadipscing elitr, sed diam nonumy eirmod</p>
    </div>
  </div>
  <div class=\\"panel-body clearfix\\">
    <div class=\\"comment-authoring\\"><strong>Bob Smith</strong> entry.authored_date</div>
    <div class=\\"pull-right comment-actions\\"><button class=\\"btn btn-sm btn-default\\" title=\\"entry.edit_comment\\"><i class=\\"fas fa-pencil\\"></i></button><button class=\\"btn btn-sm btn-danger\\" title=\\"entry.delete_comment\\"><i class=\\"fas fa-trash-alt\\"></i></button></div>
    <div class=\\"comment-content\\">
      <!-- eslint-disable vue/no-v-html -->
      <p>Et dolore magna aliquyam erat, sed diam voluptua.</p>
    </div>
  </div>
  <div class=\\"panel-body\\">
    <!--v-if-->
    <!--v-if--><button class=\\"btn btn-sm btn-primary pull-right\\"><i class=\\"fas fa-plus\\"></i> thread.add_comment</button>
  </div>
</div>"
`;
